include/ginkgo/core/solver/multigrid.hpp

```diff
@@ -0,0 +1,417 @@
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Criterion factories.
+         */
+        std::vector<std::shared_ptr<const stop::CriterionFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(criteria, nullptr);
+
+        /**
+         * MultigridLevel Factory list
+         */
+        std::vector<std::shared_ptr<const gko::LinOpFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(mg_level, nullptr);
+
+        /**
+         * Custom selector (level, matrix)
+         * default selector: use the first factory when mg_level size = 1
+         *                   use the level as the index when mg_level size > 1
+         */
+        std::function<size_type(const size_type, const LinOp *)>
+            GKO_FACTORY_PARAMETER_SCALAR(mg_level_index, nullptr);
+
+        /**
+         * Pre-smooth Factory list.
+         * Its size must be 0, 1 or be the same as mg_level's.
+         * when size = 0, do not use pre_smoother
+         * when size = 1, use the first factory
+         * when size > 1, use the same selector as mg_level
+         * nullptr skips this pre_smoother at the level, which is different from
+         * Identity Factory. Identity Factory updates x = x + relaxation *
+         * residual.
+         */
+        std::vector<std::shared_ptr<const LinOpFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(pre_smoother, nullptr);
+
+        /**
+         * Post-smooth Factory list.
+         * It is similar to Pre-smooth Factory list. It is ignored if
+         * post_uses_pre = true.
+         */
+        std::vector<std::shared_ptr<const LinOpFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(post_smoother, nullptr);
+
+        /**
+         * Mid-smooth Factory list. If it contains availble elements, multigrid
+         * always generate the corresponding smoother. However, it only involve
+         * in the procedure when cycle is k or f.
+         * It is similar to Pre-smooth Factory list. It is ignored if
+         * multigrid_mid_uses is not mid.
+         */
+        std::vector<std::shared_ptr<const LinOpFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(mid_smoother, nullptr);
+
+        /**
+         * Whether Post-related calls use corresponding pre-related calls.
+         */
+        bool GKO_FACTORY_PARAMETER_SCALAR(post_uses_pre, true);
+
+        /**
+         * Which Mid-related calls use pre/mid/post-related calls.
+         * Availble options: pre/mid/post.
+         */
+        multigrid_mid_uses GKO_FACTORY_PARAMETER_SCALAR(
+            mid_case, multigrid_mid_uses::pre);
+
+        /**
+         * The maximum level can be generated
+         */
+        size_type GKO_FACTORY_PARAMETER_SCALAR(max_levels, 10);
+
+        /**
+         * The minimal coarse rows.
+         * If generation gets the matrix which contains less than
+         * `min_coarse_rows`, the generation stops.
+         */
+        size_type GKO_FACTORY_PARAMETER_SCALAR(min_coarse_rows, 2);
```

Comment:
We should set a better default. Maybe
```suggestion
        size_type GKO_FACTORY_PARAMETER_SCALAR(min_coarse_rows, 64);
```

Reply:
yes, I forget to change it back

Reply:
@pratikvn I don't think having a coarse grid of just 2 DOFs would be ideal. Many of the lower levels would then be heavily latency-dominated. We should choose the largest possible size that something like unpreconditioned CG or jacobi CG can effectively solve quickly.

Reply:
AMGX reference uses 2 in this option.
But I think 64 or even larger makes more sense.
Use 64 now